File: backend/accounts/permissions.py
"""
Role-Based Access Control (RBAC) for the panel.

Roles (see accounts.models.CustomUser.ROLE_*):
- SUPER_ADMIN (super_admin): Full access to all endpoints.
- MANAGEMENT (management): Instagram Hub, Telegram Channels, Price Hub (read/write),
  Finalize, Analysis. No access to User Management or Site Settings.
- EMPLOYEE (employee): Read-only access to Price Hub and Dashboard (Finalize dashboard,
  Analysis). No access to Telegram Setup, Instagram Hub configuration, or Finalizing prices.

Permission mapping:
- User Management (list/create/update users, force logout, activity logs): IsSuperAdmin
- Site Settings (site settings, bots, channels, logs): IsSuperAdmin
- Instagram Hub (preview, status, config): IsSuperAdminOrManagement
- Finalize (dashboard read): IsAuthenticated; (category/special/all finalize): IsSuperAdminOrManagement
- Price Hub (list/detail/history): IsAuthenticated; (update/bulk-update): IsSuperAdminOrManagement
- Telegram app (channels, send message, bots, auto-post): IsSuperAdminOrManagementOrEmployee
  (allows super_admin, management, and employee to CRUD bots and channels)
- Analysis dashboard: IsAuthenticated

Frontend permission config: frontend/src/config/permissions.js
"""
import logging

from rest_framework import permissions

logger = logging.getLogger(__name__)


class IsSuperAdmin(permissions.BasePermission):
    """
    Allows access only to users with role super_admin.
    Used for User Management API and Site Settings.
    """

    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            logger.info(
                "IsSuperAdmin: denied (user=%s, authenticated=%s)",
                getattr(request.user, "username", None) if request.user else None,
                getattr(request.user, "is_authenticated", False) if request.user else False,
            )
            return False
        role = getattr(request.user, "role", None)
        allowed = role == "super_admin"
        logger.info(
            "IsSuperAdmin: user=%s role=%r allowed=%s",
            request.user.username,
            role,
            allowed,
        )
        return allowed


class IsSuperAdminOrManagement(permissions.BasePermission):
    """
    Allows access only to users with role super_admin or management.
    Used for: Instagram Hub, Telegram setup/channels, Finalize actions,
    Price Hub write (update/bulk-update), Template editor.
    """

    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            logger.info(
                "IsSuperAdminOrManagement: denied (user=%s, authenticated=%s)",
                getattr(request.user, "username", None) if request.user else None,
                getattr(request.user, "is_authenticated", False) if request.user else False,
            )
            return False
        role = getattr(request.user, "role", None)
        allowed = role in ("super_admin", "management")
        logger.info(
            "IsSuperAdminOrManagement: user=%s role=%r allowed=%s",
            request.user.username,
            role,
            allowed,
        )
        return allowed


class IsSuperAdminOrManagementOrEmployee(permissions.BasePermission):
    """
    Allows access for users with role super_admin, management, or employee.
    Used for: Telegram bots and channels CRUD (and related telegram hub actions)
    so that employees can manage bots/channels and send messages.
    """

    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            logger.info(
                "IsSuperAdminOrManagementOrEmployee: denied (user=%s, authenticated=%s)",
                getattr(request.user, "username", None) if request.user else None,
                getattr(request.user, "is_authenticated", False) if request.user else False,
            )
            return False
        if getattr(request.user, "is_superuser", False):
            return True
        role = getattr(request.user, "role", None)
        allowed = role in ("super_admin", "management", "employee")
        logger.info(
            "IsSuperAdminOrManagementOrEmployee: user=%s role=%r allowed=%s",
            request.user.username,
            role,
            allowed,
        )
        return allowed

    def has_object_permission(self, request, view, obj):
        """Allow same as request-level: super_admin, management, employee can act on any object."""
        return self.has_permission(request, view)

File: frontend/src/config/permissions.js
/**
 * Central permission config — single source of truth for roles and access.
 *
 * MUST stay in sync with backend/accounts/permissions.py. This file only decides what the
 * UI *shows*; the API is the real gate. When the two disagree the user gets a nav item
 * that 403s on click, which is exactly what used to happen here:
 *   - `settings` was ALL_PANEL_ROLES, but SiteSettings/bots/channels/logs are IsSuperAdmin.
 *   - `adminManagement` allowed `management`, but User Management is IsSuperAdmin.
 * Both were narrowed to match. `settings` has since been widened again to
 * super_admin + management — a self-serve signup owns its workspace with
 * role=management and has to be able to configure branding, uploads, fonts and
 * the webhook — and the backend was widened with it (SiteSettings and the
 * upload policy now accept IsSuperAdminOrManagement). The pieces that stayed
 * IsSuperAdmin — bots, channels and the activity log — are hidden inside the
 * page via `settingsAdmin` rather than by closing the whole page again.
 */

export const ROLES = {
  SUPER_ADMIN: 'super_admin',
  MANAGEMENT: 'management',
  DEVELOPER: 'developer',
  EMPLOYEE: 'employee',
}

/** All roles that can use the panel. */
export const ALL_PANEL_ROLES = Object.values(ROLES)

/** Backend: IsSuperAdmin — user management and site settings. */
const SUPER_ADMIN_ONLY = [ROLES.SUPER_ADMIN]

/** Backend: IsSuperAdminOrManagement — finalize, price writes, template editor, Instagram. */
const SUPER_ADMIN_OR_MANAGEMENT = [ROLES.SUPER_ADMIN, ROLES.MANAGEMENT]

/** Backend: IsSuperAdminOrManagementOrEmployee — Telegram bots/channels CRUD and sending. */
const SUPER_ADMIN_OR_MANAGEMENT_OR_EMPLOYEE = [ROLES.SUPER_ADMIN, ROLES.MANAGEMENT, ROLES.EMPLOYEE]

/**
 * Which roles can access each feature. Mirrors backend/accounts/permissions.py.
 *
 * NOTE on ROLES.DEVELOPER: it is a valid choice on CustomUser.ROLE_CHOICES but no
 * backend permission class accepts it — not IsSuperAdmin, not IsSuperAdminOrManagement,
 * not IsSuperAdminOrManagementOrEmployee. A developer therefore only reaches the plain
 * IsAuthenticated endpoints (analysis, finalize dashboard read, price hub read). It is
 * listed here only where that is actually true, instead of being shown a panel that 403s.
 */
export const PERMISSIONS = {
  /** تحلیل و نمودارها — IsAuthenticated (همه) */
  analysis: ALL_PANEL_ROLES,
  /** نهایی‌سازی و انتشار به تلگرام — IsSuperAdminOrManagement */
  finalize: SUPER_ADMIN_OR_MANAGEMENT,
  /** تنظیمات پنل (برندینگ، آپلود، فونت، وبهوک) — IsSuperAdminOrManagement */
  settings: SUPER_ADMIN_OR_MANAGEMENT,
  /** بخش‌های سراسری تنظیمات (ربات، کانال، لاگ‌ها) — IsSuperAdmin */
  settingsAdmin: SUPER_ADMIN_ONLY,
  /** ربات و کانال تلگرام — IsSuperAdminOrManagementOrEmployee */
  telegram: SUPER_ADMIN_OR_MANAGEMENT_OR_EMPLOYEE,
  /** صف سفارش‌های واتس‌اپ و مینی‌اپ — IsSuperAdminOrManagementOrEmployee */
  orders: SUPER_ADMIN_OR_MANAGEMENT_OR_EMPLOYEE,
  /** حذف آیتم‌ها (دسته‌بندی، قالب و غیره) — IsSuperAdminOrManagement */
  deleteItems: SUPER_ADMIN_OR_MANAGEMENT,
  /** مدیریت کاربران / ادمین‌ها (اضافه، ویرایش، حذف، لاگ فعالیت) — IsSuperAdmin */
  adminManagement: SUPER_ADMIN_ONLY,
  programmerHub: [ROLES.DEVELOPER, ROLES.SUPER_ADMIN],
}

function normalizeRole(role) {
  const key = String(role ?? '').trim().toLowerCase().replaceAll('-', '_')
  if (key === 'superadmin') return ROLES.SUPER_ADMIN
  return key
}

/**
 * @param {string|null} role
 * @param {string} permission - key of PERMISSIONS
 * @returns {boolean}
 */
export function can(role, permission) {
  const normalizedRole = normalizeRole(role)
  if (!normalizedRole) return false
  const allowed = PERMISSIONS[permission]
  return Array.isArray(allowed) && allowed.includes(normalizedRole)
}

export default { ROLES, PERMISSIONS, ALL_PANEL_ROLES, can }
